File: tree/benchmarks/heart_disease.py
from pandas.api.types import CategoricalDtype
from sklearn.metrics import accuracy_score, roc_auc_score
from sklearn.model_selection import train_test_split
import argparse
import numpy as np
import pandas as pd
import json

parser = argparse.ArgumentParser()
parser.add_argument('--library', choices=['h2o', 'lightgbm', 'sklearn', 'xgboost', 'catboost'], required=True)
args = parser.parse_args()

# Load the data.
path_train = 'data/heart_disease_train.csv'
path_test = 'data/heart_disease_test.csv'
target_column_name = "diagnosis"
gender_options = ['male', 'female']
chest_pain_options = ['typical angina', 'asymptomatic', 'non-angina pain', 'atypical angina']
fasting_blood_sugar_greater_than_120_options = [True, False]
resting_ecg_result_options = ['probable or definite left ventricular hypertrophy', 'normal', 'ST-T wave abnormality']
exercise_induced_angina_options = ['no', 'yes']
exercise_st_slope_options = ['downsloping', 'flat', 'upsloping']
fluoroscopy_vessels_colored_options = ['0', '1', '2', '3']
thallium_stress_test_options = ['fixed defect', 'normal', 'reversible defect']
diagnosis_options = ['Negative', 'Positive']
dtype = {
	'age': np.float64,
	'gender': CategoricalDtype(categories=gender_options),
	'chest_pain': CategoricalDtype(categories=chest_pain_options),
	'resting_blood_pressure': np.float64,
	'cholesterol': np.float64,
	'fasting_blood_sugar_greater_than_120': CategoricalDtype(categories=fasting_blood_sugar_greater_than_120_options),
	'resting_ecg_result': CategoricalDtype(categories=resting_ecg_result_options),
	'exercise_max_heart_rate': np.float64,
	'exercise_induced_angina': CategoricalDtype(categories=exercise_induced_angina_options),
	'exercise_st_depression': np.float64,
	'exercise_st_slope': CategoricalDtype(categories=exercise_st_slope_options),
	'fluoroscopy_vessels_colored': CategoricalDtype(categories=fluoroscopy_vessels_colored_options),
	'thallium_stress_test': CategoricalDtype(categories=thallium_stress_test_options),
	'diagnosis': CategoricalDtype(categories=diagnosis_options)
}
data_train = pd.read_csv(path_train, dtype=dtype)
data_test = pd.read_csv(path_test, dtype=dtype)
if args.library == 'xgboost' or args.library == 'sklearn' or args.library == 'catboost':
	categorical_columns = data_train.select_dtypes(['category']).columns
	data_train.loc[:, categorical_columns] = data_train.loc[:, categorical_columns].apply(lambda x: x.cat.codes)
	data_test.loc[:, categorical_columns] = data_test.loc[:, categorical_columns].apply(lambda x: x.cat.codes)
features_train = data_train.loc[:, data_train.columns != target_column_name]
labels_train = data_train[target_column_name]
features_test = data_test.loc[:, data_test.columns != target_column_name]
labels_test = data_test[target_column_name]

# Train the model.
if args.library == 'h2o':
	import h2o
	from h2o.estimators import H2OGradientBoostingEstimator
	h2o.init()
	h2o.no_progress()
	data_train = pd.concat([features_train, labels_train], axis=1)
	data_test = pd.concat([features_test, labels_test], axis=1)
	data_train = h2o.H2OFrame(python_obj=data_train)
	data_test = h2o.H2OFrame(python_obj=data_test)
	feature_column_names = [column for column in data_train.columns if column != target_column_name]
	model = H2OGradientBoostingEstimator(
		distribution="bernoulli",
		learn_rate=0.1,
		nbins=255,
		ntrees=100,
	)
	model.train(
		training_frame=data_train,
		y=target_column_name,
		x=feature_column_names,
	)
elif args.library == 'lightgbm':
	import lightgbm as lgb
	model = lgb.LGBMClassifier(
		learning_rate=0.1,
		n_estimators=100,
		num_leaves=255,
	)
	model.fit(features_train, labels_train, )
elif args.library == 'sklearn':
	from sklearn.experimental import enable_hist_gradient_boosting
	from sklearn.ensemble import HistGradientBoostingClassifier
	model = HistGradientBoostingClassifier(
		learning_rate=0.1,
		max_iter=100,
		max_leaf_nodes=255,
		validation_fraction=None,
	)
	model.fit(features_train, labels_train)
elif args.library == 'xgboost':
	import xgboost as xgb
	model = xgb.XGBClassifier(
		eta=0.1,
		grow_policy='lossguide',
		n_estimators=100,
		tree_method='hist',
	)
	model.fit(features_train, labels_train)
elif args.library == 'catboost':
	from catboost import CatBoostClassifier
	categorical_columns = [column for column in categorical_columns if column != target_column_name]
	model = CatBoostClassifier(
		cat_features=categorical_columns,
		grow_policy='Lossguide',
		learning_rate=0.1,
		n_estimators=100,
		num_leaves=255,
		train_dir='data/catboost_info',
		verbose=False
	)
	model.fit(features_train, labels_train, silent=True)

# Make predictions on the test data.
if args.library == 'h2o':
	predictions_proba = model.predict(data_test).as_data_frame()['Positive']
else:
	predictions_proba = model.predict_proba(features_test)[:, 1]

# Compute metrics.
auc_roc = roc_auc_score(labels_test, predictions_proba)

# Compute memory usage.
f = open("/proc/self/status", "r")
for line in f.readlines():
	if line.startswith("VmHWM"):
		memory = line.split(":")[1].strip()

print(json.dumps({
	'auc_roc': auc_roc,
	'memory': memory,
}))

File: crates/tree/benchmarks/heart_disease.rs
use modelfox_table::prelude::*;
use modelfox_tree::Progress;
use modelfox_zip::zip;
use ndarray::prelude::*;
use serde_json::json;
use std::{collections::BTreeMap, path::Path};

fn main() {
	// Load the data.
	let csv_file_path_train = Path::new("data/heart_disease_train.csv");
	let csv_file_path_test = Path::new("data/heart_disease_test.csv");
	let _n_rows_train = 242;
	let n_rows_test = 61;
	let target_column_index = 13;
	let gender_variants = ["male", "female"].iter().map(ToString::to_string).collect();
	let chest_pain_variants = [
		"typical angina",
		"asymptomatic",
		"non-angina pain",
		"atypical angina",
	]
	.iter()
	.map(ToString::to_string)
	.collect();
	let fasting_blood_sugar_greater_than_120_variants =
		["True", "False"].iter().map(ToString::to_string).collect();
	let resting_ecg_result_variants = [
		"probable or definite left ventricular hypertrophy",
		"normal",
		"ST-T wave abnormality",
	]
	.iter()
	.map(ToString::to_string)
	.collect();
	let exercise_induced_angina_variants = ["no", "yes"].iter().map(ToString::to_string).collect();
	let exercise_st_slope_variants = ["downsloping", "flat", "upsloping"]
		.iter()
		.map(ToString::to_string)
		.collect();
	let thallium_stress_test_variants = ["fixed defect", "normal", "reversible defect"]
		.iter()
		.map(ToString::to_string)
		.collect();
	let fluoroscopy_vessels_colored_variants = ["0", "1", "2", "3"]
		.iter()
		.map(ToString::to_string)
		.collect();
	let diagnosis_variants = ["Negative", "Positive"]
		.iter()
		.map(ToString::to_string)
		.collect();
	let options = modelfox_table::Options {
		column_types: Some(BTreeMap::from([
			("age".to_owned(), TableColumnType::Number),
			(
				"gender".to_owned(),
				TableColumnType::Enum {
					variants: gender_variants,
				},
			),
			(
				"chest_pain".to_owned(),
				TableColumnType::Enum {
					variants: chest_pain_variants,
				},
			),
			("resting_blood_pressure".to_owned(), TableColumnType::Number),
			("cholesterol".to_owned(), TableColumnType::Number),
			(
				"fasting_blood_sugar_greater_than_120".to_owned(),
				TableColumnType::Enum {
					variants: fasting_blood_sugar_greater_than_120_variants,
				},
			),
			(
				"resting_ecg_result".to_owned(),
				TableColumnType::Enum {
					variants: resting_ecg_result_variants,
				},
			),
			(
				"exercise_max_heart_rate".to_owned(),
				TableColumnType::Number,
			),
			(
				"exercise_induced_angina".to_owned(),
				TableColumnType::Enum {
					variants: exercise_induced_angina_variants,
				},
			),
			("exercise_st_depression".to_owned(), TableColumnType::Number),
			(
				"exercise_st_slope".to_owned(),
				TableColumnType::Enum {
					variants: exercise_st_slope_variants,
				},
			),
			(
				"fluoroscopy_vessels_colored".to_owned(),
				TableColumnType::Enum {
					variants: fluoroscopy_vessels_colored_variants,
				},
			),
			(
				"thallium_stress_test".to_owned(),
				TableColumnType::Enum {
					variants: thallium_stress_test_variants,
				},
			),
			(
				"diagnosis".to_owned(),
				TableColumnType::Enum {
					variants: diagnosis_variants,
				},
			),
		])),
		..Default::default()
	};
	let mut features_train =
		Table::from_path(csv_file_path_train, options.clone(), &mut |_| {}).unwrap();
	let labels_train = features_train.columns_mut().remove(target_column_index);
	let mut features_test =
		Table::from_path(csv_file_path_test, options.clone(), &mut |_| {}).unwrap();
	let labels_test = features_test.columns_mut().remove(target_column_index);
	let labels_train = labels_train.as_enum().unwrap();
	let labels_test = labels_test.as_enum().unwrap();

	// Train the model.
	let train_output = modelfox_tree::BinaryClassifier::train(
		features_train.view(),
		labels_train.view(),
		&modelfox_tree::TrainOptions {
			max_leaf_nodes: 255,
			..Default::default()
		},
		Progress {
			kill_chip: &modelfox_kill_chip::KillChip::default(),
			handle_progress_event: &mut |_| {},
		},
	);

	// Make predictions on the test data.
	let features_test = features_test.to_rows();
	let mut probabilities = Array::zeros(n_rows_test);
	train_output
		.model
		.predict(features_test.view(), probabilities.view_mut());

	// Compute metrics.
	let input = zip!(probabilities.iter(), labels_test.iter())
		.map(|(probability, label)| (*probability, label.unwrap()))
		.collect();
	let auc_roc = modelfox_metrics::AucRoc::compute(input);

	let output = json!({
		"auc_roc": auc_roc,
	});
	println!("{}", output);
}
